libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/plugin/KotlinPlugin.kt

```diff
@@ -734,18 +734,32 @@ abstract class AbstractAndroidProjectHandler(private val kotlinConfigurationTool
         val project = kotlinAndroidTarget.project
         val ext = project.extensions.getByName("android") as BaseExtension
 
-        ext.sourceSets.all { sourceSet ->
-            logger.kotlinDebug("Creating KotlinBaseSourceSet for source set $sourceSet")
-            val kotlinSourceSet = project.kotlinExtension.sourceSets.maybeCreate(
-                lowerCamelCaseName(kotlinAndroidTarget.disambiguationClassifier, sourceSet.name)
-            ).apply {
-                kotlin.srcDir(project.file(project.file("src/${sourceSet.name}/kotlin")))
-                kotlin.srcDirs(sourceSet.java.srcDirs)
+        fun kotlinSourceSetName(sourceSet: AndroidSourceSet): String {
+            return lowerCamelCaseName(kotlinAndroidTarget.disambiguationClassifier, sourceSet.name)
+        }
+
+        val kotlinSourceSetNames = ext.sourceSets.map(::kotlinSourceSetName)
+
+        ext.sourceSets.all { androidSourceSet ->
+            logger.kotlinDebug("Creating KotlinBaseSourceSet for source set $androidSourceSet")
+            val kotlinSourceSet = project.kotlinExtension.sourceSets.maybeCreate(kotlinSourceSetName(androidSourceSet)).apply {
+
+                // Avoid overlapping source directories like src/androidTest/kotlin which would otherwise be present  in the
+                // androidTest and androidAndroidTest kotlin source sets
+                if (kotlinAndroidTarget.disambiguationClassifier == null || androidSourceSet.name !in kotlinSourceSetNames) {
```

Comment:
@h0tk3y Is this considered a breaking change?

Reply:
Because I think it is not. The source directory being present in both kotlin source sets already enforced that the sources in the directory were compatible for both source sets. Removing the overlap will fix the unwanted behavior while it should not lead to code not compiling (that did compile beforehand)